File: packages/core/src/core/report.js
import { writeFile } from "fs/promises";
import { join, relative } from "path";
import { stringify } from "csv-stringify/sync";

/**
 * Writes a per-file CSV report to the destination directory.
 *
 * @param {Array} manifest - The full manifest from the pipeline
 * @param {Object} options - Pipeline options
 * @param {string} options.source - Source directory path
 * @param {string} options.destination - Destination directory path
 * @returns {Promise<string>} The path to the written report file
 */
export async function report(manifest, options = {}) {
  const { logger } = options;

  const rows = manifest.map((entry) => {
    // All paths are relative to the source directory
    const media = relative(options.source, entry.mediaPath);

    // Empty if no sidecar was matched
    const metadata = entry.json?.path ? relative(options.source, entry.json.path) : "";

    // Empty if link stage did not assign a confidence level
    const confidence = entry.json?.confidence ?? "";

    const { status, reason } = resolveStatus(entry);

    return [media, metadata, String(confidence), status, reason];
  });

  const reportPath = join(options.destination, "report.csv");

  await writeFile(
    reportPath,
    stringify(rows, {
      header: true,
      columns: ["media", "metadata", "confidence", "status", "reason"],
    })
  );

  logger?.info(`Report written to ${reportPath}`);
  return reportPath;
}

// Walks the entry's stage properties to find its terminal status.
function resolveStatus(entry) {
  // Check apply stage first (terminal state for processed files)
  if (entry.apply?.status === "embedded") {
    return {
      status: "embedded",
      reason: "",
    };
  }

  if (entry.apply?.status === "copied") {
    return {
      status: "copied",
      reason: "",
    };
  }

  if (entry.apply?.status === "error") {
    return {
      status: "error",
      reason: entry.apply.reason ?? "",
    };
  }

  // Check reconcile stage
  if (entry.metadata?.status === "skipped") {
    return {
      status: "skipped",
      reason: entry.metadata.reason ?? "",
    };
  }

  // Check dedupe stage
  if (entry.dedupe?.status === "delete") {
    return {
      status: "deleted",
      reason: "",
    };
  }

  if (entry.dedupe?.status === "error") {
    return {
      status: "error",
      reason: entry.dedupe.reason ?? "",
    };
  }

  return {
    status: "error",
    reason: "Unknown state",
  };
}

File: packages/core/src/pixelkasten/core/report.py
"""
CSV report generation — summarize pipeline results per file.

Ported from packages/core/src/core/report.js. Uses Python stdlib csv module.
"""

import csv
import os

from pixelkasten.core.options import PipelineOptions


def report(manifest: list[dict], options: PipelineOptions) -> str:
    """
    Write a per-file CSV report to the destination directory.

    Columns: media, metadata, confidence, status, reason
    All paths are relative to the source directory.

    Returns the path to the written report file.
    """
    source = options.source
    destination = options.destination or ""

    report_path = os.path.join(destination, "report.csv")

    with open(report_path, "w", newline="") as f:
        writer = csv.writer(f)
        writer.writerow(["media", "metadata", "confidence", "status", "reason"])

        for entry in manifest:
            # Use forward slashes in CSV output (portable, matches Node.js behavior)
            media = os.path.relpath(entry["mediaPath"], source).replace("\\", "/")

            metadata = ""
            json_info = entry.get("json")
            if json_info and json_info.get("path"):
                metadata = os.path.relpath(json_info["path"], source).replace("\\", "/")

            confidence = ""
            if json_info and json_info.get("confidence") is not None:
                confidence = str(json_info["confidence"])

            status_info = resolve_status(entry)

            writer.writerow(
                [
                    media,
                    metadata,
                    confidence,
                    status_info["status"],
                    status_info["reason"],
                ]
            )

    return report_path


def resolve_status(entry: dict) -> dict:
    """
    Walk the entry's stage properties to find its terminal status.

    Priority order:
    1. apply.status == "embedded"
    2. apply.status == "copied"
    3. apply.status == "error"
    4. metadata.status == "skipped"
    5. dedupe.status == "delete"
    6. dedupe.status == "error"
    7. Fallback: error/unknown
    """
    apply_status = entry.get("apply", {}).get("status")

    if apply_status == "embedded":
        return {"status": "embedded", "reason": ""}

    if apply_status == "copied":
        return {"status": "copied", "reason": ""}

    if apply_status == "error":
        return {"status": "error", "reason": entry.get("apply", {}).get("reason", "")}

    if entry.get("metadata", {}).get("status") == "skipped":
        return {
            "status": "skipped",
            "reason": entry.get("metadata", {}).get("reason", ""),
        }

    dedupe_status = entry.get("dedupe", {}).get("status")

    if dedupe_status == "delete":
        return {"status": "deleted", "reason": ""}

    if dedupe_status == "error":
        return {
            "status": "error",
            "reason": entry.get("dedupe", {}).get("reason", ""),
        }

    return {"status": "error", "reason": "Unknown state"}
